pkg/action/hooks.go

```diff
@@ -83,6 +86,19 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
 		err = cfg.KubeClient.WatchUntilReady(resources, timeout)
 		// Note the time of success/failure
 		h.LastRun.CompletedAt = helmtime.Now()
+
+		if isTestHook(h) {
+			client, err := cfg.KubernetesClientSet() // "Why create a new client for each hook?" Because `cfg.RESTClientGetter` can be `nil` before the `for` loop, as detected by the `TestInstall/basic_install` and `TestInstallRelease` tests.
```

Comment:
I'm really confused. Why are we creating a new client here? Why is there no check for this earlier? On line 86 we call `cfg.KubeClient` to perform a WatchUntilReady. In other words the argument being made here seems illogical because we've already used the client prior to this line, so it shouldn't be nil.

Reply:
> Why are we creating a new client _here_ [and not somewhere else]?

If we move `client, err := cfg.KubernetesClientSet()` to _before_ the `for _, h := range executingHooks` loop, then the `TestInstall/basic_install` and `TestInstallRelease` tests fail because `cfg.RESTClientGetter` is `nil` when `cfg.KubernetesClientSet` is called.

> Why are we creating a new client [at all]?

`getHookLog` must be passed a `kubernetes.Interface`, and there is none in scope.

> On line 86 we call `cfg.KubeClient` to perform a WatchUntilReady. In other words the argument being made here seems illogical because we've already used the client prior to this line, so it shouldn't be nil.

I'm also really confused. :brain: We _create_ the client on that line (`client, err := cfg.KubernetesClientSet()`), so we cannot possibly have used it prior to it, can we? On line 86, `cfg.KubeClient.WatchUntilReady` is called, but `cfg.KubeClient` is a `kube.Interface`, whereas `client` is a `kubernetes.Interface`. There seems to be some misunderstanding going on here.

One might argue that `getHookLog` should be rewritten so that it doesn't require a `kubernetes.Interface`, but I have no idea how to do so. We probably just copied that code from somewhere else in the code base and modified it for our purpose.

Reply:
I'm finally starting to understand the problem at hand. The `RESTClientGetter` field in the config is assigned a non-`nil` value when actions are run as part of actual commands, but not when they are tested. So any code using `KubernetesClientSet` will probably panic during testing, since [it assumes the `RESTClientGetter` field to have a non-`nil` value](https://github.com/helm/helm/blob/8ca401398d8b6dfd41b46f446abcd5285e219829/pkg/action/action.go#L281).

Reply:
This probably implies that we cannot test [`GetPodLogs`](https://github.com/helm/helm/blob/8ca401398d8b6dfd41b46f446abcd5285e219829/pkg/action/release_testing.go#L104) (which calls the `KubernetesClientSet` method) without refactoring it substantially.

Reply:
We managed to populate the `RESTClientGetter` field in the affected test cases with `cli.New().RESTClientGetter()`. However, that `RESTClientGetter` is only useful to communicate with an actual cluster, not when working with fake releases in a testing scenario. All in all, it seems like the tests requested for this PR are inherently complicated to write.

Reply:
I'd say the most problematic line from the perspective of writing tests is [this one](https://github.com/helm/helm/blob/8ca401398d8b6dfd41b46f446abcd5285e219829/pkg/action/release_testing.go#L113):

```go
req := client.CoreV1().Pods(r.Namespace).GetLogs(h.Name, &v1.PodLogOptions{})
```

I find it non-trivial to see how this could be mocked in the test suite.

Reply:
It would be wonderful if @bacongobbler or someone else could help us out with this part. What has the approach been in other cases when Helm have wanted to write tests for something that more-or-less requires connectivity to a cluster?

Reply:
We finally managed to piece together a unit test by adding `GetHookLog` as a field in `action.Configuration`, so that we could mock it; see e0b9ef060198e7836bde9408699254196cce9c8f or `git show e0b9ef060198e7836bde9408699254196cce9c8f --color-moved`.

Couldn't really avoid creating a new `kubernetes.Interface` for each hook though, but, as [previously stated](https://github.com/helm/helm/pull/9677#discussion_r765903140), that may not be a major issue.

Reply:
Working on a unit test for the command (in contrast to the action).

Reply:
The problem right now is that [`execHook`](https://github.com/helm/helm/blob/8ca401398d8b6dfd41b46f446abcd5285e219829/pkg/action/hooks.go#L30) overwrites the `StartedAt` and `CompletedAt` fields with `helmtime.Now()`, whose value can't be predicted and included in a golden-output text file.

Reply:
👆 Almost done with a solution to that problem.

Reply:
I have added unit tests for the command now!